### PR TITLE
ReactiveSocket interface

### PR DESCRIPTION
**Pull request by yschimke**
Clients wanting to mock ReactiveSocket face a tough ask without an interface.  While technically possible its not as elegant and the other option is mocking DuplexConnection or using Inline connection, both of which are much lower level abstractions.

### CODE REVIEW

File: src/ReactiveSocket.h [hunk 1]
```diff
@@ -69,53 +57,53 @@ class ReactiveSocket {
       std::unique_ptr<RequestHandlerBase> handler,
       Stats& stats = Stats::noop());
 
-  std::shared_ptr<Subscriber<Payload>> requestChannel(
+  virtual std::shared_ptr<Subscriber<Payload>> requestChannel(
```
Comment (yschimke) [inline src/ReactiveSocket.h]: I'll remove the virtuals as override is already there

**Comment by lexs** [comment]
It's problematic because clients (eg ConnectionManager) will call `fromClientConnection()` et all which are static methods. I'm not sure how you'd make that mockable?

**Comment by yschimke** [comment]
@lexs my full diff has a factory which actually works nicely and allows more flexibility, it also removes concerns like the KeepAliveTimer from the ConnectionManager

class ReactiveSocketFactory {
 public:
  virtual ~ReactiveSocketFactory() = default;
  virtual std::unique_ptr<::reactivesocket::ReactiveSocketInterface> create(
      std::unique_ptr<::reactivesocket::DuplexConnection> connection,
      ::reactivesocket::ResumeIdentificationToken resumeToken) = 0;
};

**Comment by yschimke** [comment]
But generally even assuming that there are issues for some callers, there will be some code that is given a ReactiveSocket and should be able to test against that interface.

This stuff is trivial in java land e.g. https://github.com/ReactiveSocket/reactivesocket-java/blob/7f3cf71fcd616442a68e151da6629a216632ce13/reactivesocket-core/src/test/java/io/reactivesocket/test/util/MockReactiveSocket.java

or Mockito

**Comment by lehecka** [comment]
Similarly to Java implementation, I would suggest to make all public methods of ReactiveSocket class virtual so that you can create mocks and override what you want. I don't see much value in having the extra interface. Sending ReactiveSocketInterface as parameter everywhere instead of ReactiveSocket doesn't look very pretty ;(
If you see a value in having the interface and if you plan on using it on places where the virtual methods would not be enough, then feel free to check it in.

**Comment by yschimke** [comment]
@lehecka I guess that's really my point, that's what I want.  The Java version has

`public interface ReactiveSocket {`

and impls like

```
/**
 * Server side ReactiveSocket. Receives {@link Frame}s from a
 * {@link ClientReactiveSocket}
 */
public class ServerReactiveSocket implements ReactiveSocket {

/**
 * Client Side of a ReactiveSocket socket. Sends {@link Frame}s
 * to a {@link ServerReactiveSocket}
 */
public class ClientReactiveSocket implements ReactiveSocket {
```

Plus LoadBalancer and BackupRequestSocket.

So if we want to match the Java I would drop the Interface suffix, and call the implementation class something else, e.g. StandardReactiveSocket.  I'm happy to make this change.  Would you prefer that?  I'd like to get this right.

**Comment by lehecka** [comment]
That's not a bad idea. I don't see much benefit to that, but I also don't see any harm to it. We aren't at the feature parity with Java yet so I am not sure what different implementations of the interface we will have, but maybe one day we will have more than one implementation. I will leave the decision up to you and will accept you diffs.

File: src/ReactiveSocket.h [hunk 2]
```diff
@@ -13,149 +13,65 @@
 #include "src/Stats.h"
 
 namespace folly {
-class Executor;
+  class Executor;
 }
 
 namespace reactivesocket {
+  class ClientResumeStatusCallback;
```
Comment (lehecka) [inline src/ReactiveSocket.h]: just currious, did you run ./devtools/format_all.sh?

**Pull request merged**